keps/sig-node/3085-pod-conditions-for-starting-completition-of-sandbox-creation/README.md

```diff
@@ -0,0 +1,1618 @@
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+Pod sandbox creation is a critical phase of a pod's lifecycle that the kubelet
+orchestrates across multiple components: in-tree volume plugins (ConfigMap,
+Secret, EmptyDir, etc), CSI plugins and container runtime (which in turn invokes
+a runtime handler and CNI plugins). Completion of all these phases is marked by
+the CRI runtime reporting a pod sandbox with networking configured. This KEP
+proposes a `PodHasNetwork` pod condition in pod status to indicate successful
+completion of pod sandbox creation (with networking configured) by Kubelet and
+the CRI container runtime. The `PodHasNetwork` condition will mark an important
+milestone in the pod's lifecycle similar to `ContainersReady` and the overall
+`Ready` conditions in pod status today. In the future, a dedicated
+`PodHasVolumes` condition may be considered to mark successful mounting and
+setup of all volumes in the pod spec. An alternate name like `SandboxReady` is
+avoided since Kubernetes does not directly surface low level sandbox related
+concepts to all users.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Today, the scheduler surfaces a specific pod condition: `PodScheduled` that
+clearly identifies whether a pod got scheduled by the scheduler and when
+scheduling completed. However, no specific conditions around initialization of
+successfully scheduled pods from the perspective of completion of pod sandbox
+creation (marked by the presence of a pod sandbox with networking configured) is
+surfaced to cluster administrators in a scoped and consumable fashion.
+
+There is an existing pod condition: `Initialized` that tracks execution of init
+containers. For pods without init containers, the `Initialized` condition is set
+when the Kubelet starts to process a pod before any sandbox creation activities
+start. For pods with init containers, the `Initialized` condition is set when
+init containers have been pulled and executed to completion. Therefore, the
+existing `Initialized` condition is insufficient and inaccurate for tracking
+completion of sandbox creation of all pods in a cluster. This distinction
+becomes especially relevant in multi-tenant clusters where individual tenants
+own the pod specs (including the set of init containers) while the cluster
+administrators are in charge of storage plugins, networking plugins and
+container runtime handlers.
```

Comment:
Thanks for pointing out the mismatching phase interprets on  `Initialized` between pods with init containers and without. But the issue isn't obvious to the users since we hide the pod sandbox creation completely in the implementation. To the end user, it is more like all init containers are successfully run, now it is the time to start the app containers. With the newly proposed PodCondition: PodHasNetwork, shouldn't the issue surface to the end users now?

With init containers:       PodHasNetwork -> Initialized -> ... -> Ready
Without init containers:  Initialized -> PodHasNetwork -> ... -> Ready 

Can we consolidate above flows? We can move Initialized after PodHasNetwork to indicate Kubelet now can start any app containers.

Reply:
> But the issue isn't obvious to the users since we hide the pod sandbox creation completely in the implementation. To the end user, it is more like all init containers are successfully run, now it is the time to start the app containers. With the newly proposed PodCondition: PodHasNetwork, shouldn't the issue surface to the end users now?

The inconsistency already surfaces with the way things are today when CSI/CRI/CNI is unable to mount volumes or launch the pod sandbox upon encountering errors: without init containers, today, the pod status reports the Initialized condition to be true but Kubelet also reports error events like `FailedCreatePodSandBox`, `FailedMount`, etc against the pod (which are reported as part of `kubectl describe` etc)

> Can we consolidate above flows? We can move Initialized after PodHasNetwork to indicate Kubelet now can start any app containers.

We can certainly consider that. I will add that to the KEP post-merge of this PR.

Reply:
We had a discussion on this in sig-node on July 5th. The suggestion was to carry on with the `PodHasNetwork` condition as described in this KEP as the ordering of transition of pod conditions should not be important. We can separately address the situation with `Initialized` condition for pods without init containers (mentioned above) in a separate KEP as that feels like an independent problem.